apis/python/src/tiledbsoma/indexed_dataframe.py

```diff
@@ -200,18 +199,57 @@ def read(
             )
 
+            # Acceptable ways to index:
+            # * None
+            # * A sequence of coordinates is accepted, one per dimension.
+            # * Sequence length must be at least one and <= number of dimensions.
+            # * If the sequence contains missing coordinates (length less than number of dimensions),
+            #   then "slice(None)" is assumed for the missing dimensions.
+            # * Per-dimension, explicitly specified coordinates can be one of: None, a value, a
+            #   list/ndarray/paarray/etc of values, a slice, etc.
+
             if ids is not None:
-                dim_name = A.schema.domain.dim(0).name
-                if isinstance(ids, list):
-                    sr.set_dim_points(dim_name, ids)
-                elif isinstance(ids, pa.ChunkedArray):
-                    sr.set_dim_points(dim_name, ids)
-                elif isinstance(ids, pa.Array):
-                    sr.set_dim_points(dim_name, pa.chunked_array(ids))
-                elif isinstance(ids, slice):
-                    lo_hi = util.slice_to_range(ids)
-                    if lo_hi is not None:
-                        sr.set_dim_ranges(dim_name, lo_hi)
+                if not isinstance(ids, (list, tuple)):
+                    raise TypeError(
+                        f"ids type {type(ids)} unsupported; expected list or tuple"
+                    )
+                if len(ids) < 1 or len(ids) > A.schema.domain.ndim:
+                    raise ValueError(
+                        f"ids {ids} must have length between 1 and ndim ({A.schema.domain.ndim}); got {len(ids)}"
+                    )
+
+                for i, dim_ids in enumerate(ids):
+                    # Example: ids = [None, 3, slice(4,5)]
+                    # dim_ids takes on values None, 3, and slice(4,5) in this loop body.
+                    dim_name = A.schema.domain.dim(i).name
+                    if dim_ids is None:
+                        pass  # No constraint; select all in this dimension
+                    elif isinstance(dim_ids, int):
+                        # TO DO: Support non-int index types when we have non-int index support
+                        # in libtiledbsoma's SOMAReader. See also
+                        # https://github.com/single-cell-data/TileDB-SOMA/issues/418
+                        # https://github.com/single-cell-data/TileDB-SOMA/issues/419
+                        sr.set_dim_points(dim_name, [dim_ids])
+                    elif isinstance(dim_ids, np.ndarray):
+                        if dim_ids.ndim != 1:
+                            raise ValueError(
+                                f"only 1D numpy arrays may be used to index; got {dim_ids.ndim}"
+                            )
+                        sr.set_dim_points(dim_name, dim_ids)
+                    elif isinstance(dim_ids, slice):
+                        lo_hi = util.slice_to_range(dim_ids)
+                        if lo_hi is not None:
```

Comment:
Needs clarification in docstrings -- are the slice ranges inclusive or exclusive.  Ie., like TileDB or like NumPy?   Do they support negative indexing?  Etc.